serializer_impl.hh

```diff
@@ -782,4 +782,138 @@ unknown_variant_type deserialize(Input& in, boost::type<unknown_variant_type>) {
         return unknown_variant_type{ index, std::move(v) };
     });
 }
+
+template <typename T>
+class range_deserializer {
+    utils::input_stream _in;
+    uint32_t _size;
+
+public:
+    range_deserializer(const utils::input_stream& in)
+        : _in(in)
+    {
+        _size = deserialize(_in, boost::type<uint32_t>());
+    }
+
+    // Get the number of items in the vector
+    size_t size() const noexcept {
+        return _size;
+    }
+
+    class iterator;
+
+    // Sentinel for the end() position
+    class sentinel {
+        uint32_t _idx;
+
+        sentinel(uint32_t idx) noexcept
+            : _idx(idx)
+        { }
+
+        friend class range_deserializer;
+    public:
+        ssize_t operator-(const iterator& it) const noexcept {
+            return _idx - it._idx;
+        }
+    };
+
+    // Forward iterator
+    class iterator {
+        utils::input_stream _in;
+        uint32_t _idx = 0;
+        std::optional<T> _item;
```

Comment:
Maybe better to keep in range_deserializer. Iterators are presumed cheap to copy.

Reply:
Maybe we don't need it at all. forward_iterator may only be dereferenced once, so it will never be needed to read from this member.

Reply:
s/forward_iterator/input_iterator/

Reply:
I'm not sure what you meant by not needing it at all.

Reply:
dereferencing the iterator is supposed to return a T& so it needs to store T somewhere.
If we could just return T we indeed won't need to store it in the iterator.

Reply:
https://en.cppreference.com/w/cpp/named_req/InputIterator
`++it` is allowed to invalidate all previously returned references. Any copy of previous `it` values is not even required to be comparable anymore. I think we still need to store the currently de-serialized item locally.

Reply:
Yeah, I agree. range.begin() can be called more than once so we better keep the iteration state in the iterator rather than in the range.

Reply:
But the T is not part of the iteration state.

Reply:
std::optional<T> is part of the iteration state (in this version) as it determines if the input_stream should be skipped or not in `operator++`.  If we move std::optional<T> to the range, we'll need to keep an alternative `bool consumed` state in the iterator for this purpose.

Alternatively, we can copy the input stream for the dereference operator so the iterator's input_stream won't be advanced when the item is deserialized, but then `operator++` will always have to skip the item by deserializing the size yet-again, which is wasteful.